clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -451,9 +451,16 @@ private void resetOffsets(final Set<TopicPartition> partitions) {
         long startMs = time.milliseconds();
         long remaining = timeout;
         do {
+            long heartbeatExpiryMs = client.getHeartbeatExpiryMs();
             RequestFuture<Map<TopicPartition, OffsetData>> future =
                     sendListOffsetRequests(requireTimestamps, timestampsToSearch);
-            client.poll(future, remaining);
+            while (heartbeatExpiryMs > 0 && remaining > heartbeatExpiryMs && !future.isDone()) {
+                client.poll(future, heartbeatExpiryMs);
+                remaining -= heartbeatExpiryMs;
+                heartbeatExpiryMs = client.getHeartbeatExpiryMs();
+            }
+            if (!future.succeeded() && remaining > 0) client.poll(future, remaining);
```

Comment:
Isn't this the same problem we're trying to solve? The remaining time could be much larger than the heartbeat interval.

Reply:
In the while loop, remaining time is spread out over a series of polls to ensure that pollHeartbeat() has sufficient time to be called.  If it is much larger, it shouldn't be too much of a concern.

Reply:
Hmm.. I guess you are assuming that `heartbeatExpiryMs` is never 0? Do we guarantee that?

Reply:
No, it is not guaranteed that it will never be zero. That is a problem I will need to fix.